From ed2c0ad4eee1695381e06f7accb902cbcf547420 Mon Sep 17 00:00:00 2001
From: Wim Taymans <wtaymans@redhat.com>
Date: Mon, 27 Apr 2026 10:53:44 +0200
Subject: [PATCH] spa: add spa_alloca that does overflow and limit checks

Make a function like alloca but with overflow checks and a max
allocation size.

Use this function where we can and also make sure that all alloca calls
are in some way limited.

Index: repo/spa/include/spa/utils/defs.h
===================================================================
--- repo.orig/spa/include/spa/utils/defs.h
+++ repo/spa/include/spa/utils/defs.h
@@ -457,6 +457,16 @@ struct spa_error_location {
 	_strp;								\
 })
 
+#define spa_alloca(n, size, max_size)				\
+({								\
+	void *_res = NULL; 					\
+	if ((size_t)n > (size_t)max_size / (size_t)size)	\
+		errno = ENOMEM;					\
+	else							\
+		_res = alloca((size_t)n * (size_t)size);	\
+	_res;							\
+})
+
 /**
  * \}
  */
Index: repo/spa/plugins/audioconvert/audioadapter.c
===================================================================
--- repo.orig/spa/plugins/audioconvert/audioadapter.c
+++ repo/spa/plugins/audioconvert/audioadapter.c
@@ -35,6 +35,7 @@ SPA_LOG_TOPIC_DEFINE_STATIC(log_topic, "
 
 #define MAX_PORTS	(SPA_AUDIO_MAX_CHANNELS+1)
 #define MAX_RETRY	64
+#define MAX_BLOCKS	4096
 
 /** \cond */
 
@@ -356,6 +357,7 @@ static void emit_node_info(struct impl *
 
 		if (this->info.props)
 			n_items = this->info.props->n_items;
+		n_items = SPA_MIN(n_items, 1024u);
 		items = alloca((n_items + 2) * sizeof(struct spa_dict_item));
 		for (i = 0; i < n_items; i++)
 			items[i] = this->info.props->items[i];
@@ -513,6 +515,9 @@ static int negotiate_buffers(struct impl
 
 	align = SPA_MAX(align, this->max_align);
 
+	if (blocks > MAX_BLOCKS)
+		return -ENOMEM;
+
 	datas = alloca(sizeof(struct spa_data) * blocks);
 	memset(datas, 0, sizeof(struct spa_data) * blocks);
 	aligns = alloca(sizeof(uint32_t) * blocks);
@@ -1937,11 +1942,12 @@ static int load_converter(struct impl *t
 	struct spa_dict_item *items;
 	struct spa_dict cinfo;
 	char direction[16];
-	uint32_t i;
+	uint32_t i, n_items;
 
-	items = alloca((info->n_items + 1) * sizeof(struct spa_dict_item));
+	n_items = SPA_MIN(info->n_items, 1024u);
+	items = alloca((n_items + 1) * sizeof(struct spa_dict_item));
 	cinfo = SPA_DICT(items, 0);
-	for (i = 0; i < info->n_items; i++)
+	for (i = 0; i < n_items; i++)
 		items[cinfo.n_items++] = info->items[i];
 
 	snprintf(direction, sizeof(direction), "%s",
Index: repo/src/modules/module-protocol-pulse/defs.h
===================================================================
--- repo.orig/src/modules/module-protocol-pulse/defs.h
+++ repo/src/modules/module-protocol-pulse/defs.h
@@ -34,6 +34,9 @@
 
 #define MAXLENGTH		(4u*1024*1024) /* 4MB */
 
+/* pulseaudio has a 128 char limit for this but we can allow more */
+#define MAX_NAME		1024u
+
 #define SCACHE_ENTRY_SIZE_MAX	(1024*1024*16)
 
 #define MODULE_INDEX_MASK	0xfffffffu
Index: repo/src/modules/module-protocol-pulse/message.c
===================================================================
--- repo.orig/src/modules/module-protocol-pulse/message.c
+++ repo/src/modules/module-protocol-pulse/message.c
@@ -531,10 +531,13 @@ static void add_stream_group(struct mess
 	else
 		return;
 
-	write_string(m, key);
 	l = strlen(prefix) + strlen(id) + strlen(str) + 6; /* "-by-" , ":" and \0 */
+	if (l < 0 || l > 4096)
+		return;
+
+	write_string(m, key);
 	b = alloca(l);
-	snprintf(b, l, "%s-by-%s:%s", prefix, id, str);
+	spa_scnprintf(b, l, "%s-by-%s:%s", prefix, id, str);
 	write_u32(m, l);
 	write_arbitrary(m, b, l);
 }
Index: repo/src/modules/module-protocol-pulse/pulse-server.c
===================================================================
--- repo.orig/src/modules/module-protocol-pulse/pulse-server.c
+++ repo/src/modules/module-protocol-pulse/pulse-server.c
@@ -71,6 +71,7 @@
 /* The max amount of data we send in one block when capturing. In PulseAudio this
  * size is derived from the mempool PA_MEMPOOL_SLOT_SIZE */
 #define MAX_BLOCK	(64*1024)
+#define MAX_ALLOCA_SIZE	(64*1024)
 
 #define TEMPORARY_MOVE_TIMEOUT	(SPA_NSEC_PER_SEC)
 
@@ -752,8 +753,11 @@ static int reply_create_record_stream(st
 		peer_index = peer->index;
 		if (!pw_manager_object_is_source(peer)) {
 			size_t len = (name ? strlen(name) : 5) + 10;
-			peer_name = tmp = alloca(len);
-			snprintf(tmp, len, "%s.monitor", name ? name : "sink");
+			if (len <= MAX_NAME) {
+				peer_name = tmp = alloca(len);
+				spa_scnprintf(tmp, len, "%s.monitor", name ? name : "sink");
+			} else
+				peer_name = NULL;
 		} else {
 			peer_name = name;
 		}
@@ -864,7 +868,7 @@ static void manager_added(void *data, st
 				    pw_manager_object_is_monitor(peer)) {
 					int len = strlen(peer_name) + 10;
 					char *tmp = alloca(len);
-					snprintf(tmp, len, "%s.monitor", peer_name);
+					spa_scnprintf(tmp, len, "%s.monitor", peer_name);
 					peer_name = tmp;
 				}
 				if (peer_name != NULL)
@@ -3167,7 +3171,8 @@ static int do_set_port_latency_offset(st
 		return -ENOENT;
 
 	collect_card_info(card, &card_info);
-	port_info = alloca(card_info.n_ports * sizeof(*port_info));
+	if ((port_info = spa_alloca(card_info.n_ports, sizeof(*port_info), MAX_ALLOCA_SIZE)) == NULL)
+		return -errno;
 	card_info.active_profile = SPA_ID_INVALID;
 	n_ports = collect_port_info(card, &card_info, NULL, port_info);
 
@@ -3306,7 +3311,9 @@ static int do_remove_proplist(struct cli
 	}
 
 	dict.n_items = props->dict.n_items;
-	dict.items = items = alloca(sizeof(struct spa_dict_item) * dict.n_items);
+	if ((dict.items = items = spa_alloca(dict.n_items,
+			sizeof(struct spa_dict_item), MAX_ALLOCA_SIZE)) == NULL)
+		return -errno;
 	for (i = 0; i < dict.n_items; i++) {
 		items[i].key = props->dict.items[i].key;
 		items[i].value = NULL;
@@ -3589,7 +3596,8 @@ static int fill_card_info(struct client
 		TAG_U32, card_info.n_profiles,			/* n_profiles */
 		TAG_INVALID);
 
-	profile_info = alloca(card_info.n_profiles * sizeof(*profile_info));
+	if ((profile_info = spa_alloca(card_info.n_profiles, sizeof(*profile_info), MAX_ALLOCA_SIZE)) == NULL)
+		return -errno;
 	n_profiles = collect_profile_info(o, &card_info, profile_info);
 
 	for (n = 0; n < n_profiles; n++) {
@@ -3618,7 +3626,8 @@ static int fill_card_info(struct client
 		uint32_t n_ports;
 		struct port_info *port_info, *pi;
 
-		port_info = alloca(card_info.n_ports * sizeof(*port_info));
+		if ((port_info = spa_alloca(card_info.n_ports, sizeof(*port_info), MAX_ALLOCA_SIZE)) == NULL)
+			return -errno;
 		card_info.active_profile = SPA_ID_INVALID;
 		n_ports = collect_port_info(o, &card_info, NULL, port_info);
 
@@ -3633,8 +3642,8 @@ static int fill_card_info(struct client
 
 			pi = &port_info[n];
 
-			if (pi->info && pi->n_props > 0) {
-				items = alloca(pi->n_props * sizeof(*items));
+			if (pi->info && pi->n_props > 0 &&
+			    (items = spa_alloca(pi->n_props, sizeof(*items), MAX_ALLOCA_SIZE)) != NULL) {
 				dict.items = items;
 				pdict = collect_props(pi->info, &dict);
 			}
@@ -3741,7 +3750,7 @@ static int fill_sink_info(struct client
 	if (name == NULL)
 		name = "unknown";
 
-	size = strlen(name) + 10;
+	size = SPA_MIN(strlen(name) + 10, MAX_NAME);
 	monitor_name = alloca(size);
 	if (pw_manager_object_is_source(o))
 		snprintf(monitor_name, size, "%s", name);
@@ -3822,7 +3831,8 @@ static int fill_sink_info(struct client
 		uint32_t n_ports, n;
 		struct port_info *port_info, *pi;
 
-		port_info = alloca(card_info.n_ports * sizeof(*port_info));
+		if ((port_info = spa_alloca(card_info.n_ports, sizeof(*port_info), MAX_ALLOCA_SIZE)) == NULL)
+			return -errno;
 		n_ports = collect_port_info(card, &card_info, &dev_info, port_info);
 
 		message_put(m,
@@ -3938,11 +3948,11 @@ static int fill_source_info(struct clien
 	if (name == NULL)
 		name = "unknown";
 
-	size = strlen(name) + 10;
+	size = SPA_MIN(strlen(name) + 10, MAX_NAME);
 	monitor_name = alloca(size);
 	snprintf(monitor_name, size, "%s.monitor", name);
 
-	size = strlen(desc) + 20;
+	size = SPA_MIN(strlen(desc) + 20, MAX_NAME);
 	monitor_desc = alloca(size);
 	snprintf(monitor_desc, size, "Monitor of %s", desc);
 
@@ -4018,7 +4028,8 @@ static int fill_source_info(struct clien
 		uint32_t n_ports, n;
 		struct port_info *port_info, *pi;
 
-		port_info = alloca(card_info.n_ports * sizeof(*port_info));
+		if ((port_info = spa_alloca(card_info.n_ports, sizeof(*port_info), MAX_ALLOCA_SIZE)) == NULL)
+			return -errno;
 		n_ports = collect_port_info(card, &card_info, &dev_info, port_info);
 
 		message_put(m,
Index: repo/src/pipewire/buffers.c
===================================================================
--- repo.orig/src/pipewire/buffers.c
+++ repo/src/pipewire/buffers.c
@@ -235,6 +235,9 @@ int pw_buffers_negotiate(struct pw_conte
 	if ((res = param_filter(result, &input, &output, SPA_PARAM_Meta, &b)) > 0)
 		n_params += res;
 
+	if (n_params > 4096)
+		return -EINVAL;
+
 	metas = alloca(sizeof(struct spa_meta) * n_params * 2);
 
 	n_metas = 0;
Index: repo/src/pipewire/conf.c
===================================================================
--- repo.orig/src/pipewire/conf.c
+++ repo/src/pipewire/conf.c
@@ -358,8 +358,11 @@ int pw_conf_save_state(const char *prefi
 	if ((sfd = open_write_dir(path, sizeof(path), prefix)) < 0)
 		return sfd;
 
-	tmp_name = alloca(strlen(name)+5);
-	sprintf(tmp_name, "%s.tmp", name);
+	size_t tmp_name_size = strlen(name) + 5;
+	if (tmp_name_size > PATH_MAX)
+		return -EINVAL;
+	tmp_name = alloca(tmp_name_size);
+	snprintf(tmp_name, tmp_name_size, "%s.tmp", name);
 	if ((fd = openat(sfd, tmp_name,  O_CLOEXEC | O_CREAT | O_WRONLY | O_TRUNC, 0600)) < 0) {
 		res = -errno;
 		pw_log_error("can't open file '%s': %m", tmp_name);
Index: repo/src/pipewire/impl-node.c
===================================================================
--- repo.orig/src/pipewire/impl-node.c
+++ repo/src/pipewire/impl-node.c
@@ -31,6 +31,8 @@ PW_LOG_TOPIC_EXTERN(log_node);
 
 #define DEFAULT_SYNC_TIMEOUT  ((uint64_t)(5 * SPA_NSEC_PER_SEC))
 
+#define MAX_COMMAND	(64*1024u)
+
 /** \cond */
 struct impl {
 	struct pw_impl_node this;
@@ -1949,14 +1951,17 @@ static void node_event(void *data, const
 			size_t size = SPA_POD_SIZE(&event->pod);
 
 			/* turn the event and all the arguments into a command */
-			command = alloca(size);
-			memcpy(command, event, size);
-			command->body.body.type = SPA_TYPE_COMMAND_Node;
-			command->body.body.id = SPA_NODE_COMMAND_RequestProcess;
-
-			/* send the request process to the driver but only on the
-			 * server size */
-			handle_request_process_command(node->driver_node, command);
+			if ((command = spa_alloca(1, size, MAX_COMMAND)) != NULL) {
+				memcpy(command, event, size);
+				command->body.body.type = SPA_TYPE_COMMAND_Node;
+				command->body.body.id = SPA_NODE_COMMAND_RequestProcess;
+
+				/* send the request process to the driver but only on the
+				 * server size */
+				handle_request_process_command(node->driver_node, command);
+			} else {
+				pw_log_warn("%p: ignore large command", node);
+			}
 		}
 		break;
 	default:
